CombatTagPlus/src/main/java/net/minelink/ctplus/task/TagUpdateTask.java

```diff
@@ -51,6 +51,9 @@ public void run() {
             if (!plugin.getSettings().getUntagMessage().isEmpty()) {
                 player.sendMessage(plugin.getSettings().getUntagMessage());
             }
+
+            plugin.getTagManager().untag(player.getUniqueId());
```

Comment:
> @cbarber I think it is best if we guarantee that 'PlayerUntagEvent' is guaranteed to be fired before isUntagged() returns true.

There is no isUntagged method in your repository. I'm assuming you're talking about isExpired on the Tag object. There will always be the possibility for a race condition unless you fire the event from the isExpired method itself.

> In your implementation, a external plugin could act as if a player is untagged, before the untag event happens.
By only firing the untag event once you call the untag method, you require client code to call the untag method whenever a player gets untagged (as a result of the getTag method returning null), which is error prone, and reveals details of the tagging implementation.

There are three ways a player can be untagged, time expiration, player death (calls the untag method), or third party code calling the untag method directly (say a untag console command). All of these methods behave in a predictable manor firing the new event.

What client code are you trying to hide the tagging/untagging implementation from that is prevented by firing an event?

> By firing the untag event in a getTag method, you hide the details of the tagging/untagging implementation from the client code, and guarantee that the event is fired before client code realizes that the player is untagged.

Client code calls the the tag method to change state. It stands to reason client code should call untag to change the state again. Again, mirroring the pattern that was already established prior to any changes.

> By only firing the untag event once you call the untag method, you require client code to call the untag method whenever a player gets untagged (as a result of the getTag method returning null), which is error prone, and reveals details of the tagging implementation.

Again, I see untag as a calling requirement. We're changing state within the manager (acting as a ad-hoc state machine). You do this already when the player dies.

Look, I added something useful and we're disagreeing on separation of concerns in your code. At this point, our conversation has taken up more of my time than simply maintaining my own fork by pulling your upstream, so I'm going to do that instead.